cwl/src/main/scala/cwl/ParameterContext.scala

```diff
@@ -1,27 +1,111 @@
-import wom.values.{WomMap, WomOptionalValue, WomSingleFile, WomString, WomValue}
+import cats.data.NonEmptyList
+import wom.types.WomNothingType
+import wom.values.{WomArray, WomBoolean, WomFloat, WomInteger, WomMap, WomOptionalValue, WomSingleFile, WomString, WomValue}
+import cats.syntax.traverse._
+import cats.syntax.apply._
+import cats.syntax.validated._
+import cats.instances.list._
+import common.validation.ErrorOr.ErrorOr
+import cats.data.Validated._
+import wom.callable.RuntimeEnvironment
+import wom.graph.LocalName
+
+import scala.collection.JavaConverters._
 
 object ParameterContext {
   val Empty = ParameterContext()
 }
 
-case class ParameterContext(inputs: WomValue = WomOptionalValue(WomNothingType, None),
-                            self: WomValue = WomOptionalValue(WomNothingType, None),
-                            runtime: WomValue = WomOptionalValue(WomNothingType, None)) {
-  def withInputs(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet): ParameterContext = {
-    val womValueType = WomStringType
-    copy(
-      inputs = WomMap(
-        WomMapType(WomStringType, womValueType),
-        // TODO: WOM: convert inputValues (including WomFile?) to inputs using the ioFunctionSet
-        inputValues map {
-          case (name, WomSingleFile(path)) => WomString(name) -> WomString(path)
-          case (name, womValue) => WomString(name) -> womValue
-        }
-      )
-    )
+/**
+  * This class holds values that ultimately will be bound to variables in JDK's "Nashorn" Javascript execution engine.
+  *
+  * The untyped format accomplishes two goals:
+  *   * Allows multiple types to be stored in one map
+  *   * Holds values as Nashorn expects them (a map)
+  *
+  * See {{cwl.JsUtil}} for nashorn-specific implementation details.
+  *
+  * @param inputs
+  * @param self
+  * @param runtime
+  */
+case class ParameterContext(private val inputs: Map[String, AnyRef] = Map.empty,
+                       private val self: Array[Map[String, String]] = Array.empty,
+                       private val runtime: Map[String, AnyRef] = Map.empty) {
+
+  /**
+    * see <a href="http://www.commonwl.org/v1.0/CommandLineTool.html#Runtime_environment">CWL Spec</a>.
+    */
+  def setRuntime(runtimeEnvironment: RuntimeEnvironment): ParameterContext = {
+    import runtimeEnvironment._
+
+    this.copy(runtime =  Map(
+      "outdir" -> outputPath,
+      "tmpdir" -> tempPath,
+      "cores" -> cores.toString,
+      "ram" -> ram.toString,
+      "outdirSize" -> outputPathSize.toString,
+      "tmpdirSize" -> tempPathSize.toString
+    ))
+  }
+
+  /**
+    * The public method of adding values hides our untyped internal representation and verifies that a Nashorn-compatible
+    * value exists for the womvalue being passed in.
+    *
+    * @param womMap
+    * @return a new Parameter Context with the inputs values added
+    */
+  def addInputs(womMap: Map[String, WomValue]): Either[NonEmptyList[String], ParameterContext] = {
```

Comment:
Looks like this is only an `Either` so that we can do the "we haven't added a conversion from this WomValue into Javascript yet" message.

IMO since that shouldn't be a final state for any WomValue, we should just throw for now and let this be a simple `ParameterContext` return

Reply:
Also isn't this a `Checked`?